## Jumping libraries

http://gatkforums.broadinstitute.org/gatk/discussion/6326/jumping-libraries

<p>Jumping libraries are created to bypass difficult to align/map regions, such as those containing repetitive DNA sequences.  Briefly, the DNA of interest is identified, cut into fragments either with restriction enzymes or by shearing.  The size-selected fragments are ligated to adapters for bead-capture and circularized. After bead-capture, the DNA is linearized via restriction enzymes, and can be sequenced using adapter primers facing in outward [reverse/forward (RF)] directions. These library inserts are considered jumping because the ends originate from distal genomic DNA sequences and are ligated adjacent to one another during circularization.  Potential artifacts of this method include small inserts (lacking the linearizing restriction enzyme sequence), which are inward-facing [forward/reverse (FR)] (non-jumping) read pairs. In addition, chimeras result from the paired ends falling on different chromosomes, the insert size exceeding the maximum of 100 KB, or two times the mode of the insert size for outward-facing pairs. For additional information, see <a href="http://www.wikipedia.org/wiki/Jumping_library#Paired-end_sequencing">the Wikipedia article</a>. </p>