## Genotype and Validate

http://gatkforums.broadinstitute.org/gatk/discussion/61/genotype-and-validate

<h3>Please note that this article has not been updated in a very long time and may no longer be applicable. Use at your own risk.</h3>
<hr />
<h3>Introduction</h3>
<p>Genotype and Validate is a tool to asses the quality of a technology dataset for calling SNPs and Indels given a secondary (validation) datasource. </p>
<p>The simplest scenario is when you have a VCF of hand annotated SNPs and Indels, and you want to know how well a particular technology performs calling these snps. With a dataset (BAM file) generated by the technology in test, and the hand annotated VCF, you can run GenotypeAndValidate to asses the accuracy of the calls with the new technology's dataset.</p>
<p>Another option is to validate the calls on a VCF file, using a deep coverage BAM file that you trust the calls on. The GenotypeAndValidate walker will make calls using the reads in the BAM file and take them as truth, then compare to the calls in the VCF file and produce a truth table.</p>
<h3>Command-line arguments</h3>
<p>Usage of GenotypeAndValidate and its command line arguments are described <a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_validation_GenotypeAndValidate.html">here</a>.</p>
<h3>The VCF Annotations </span></h2></h3>
<p>The annotations can be either true positive (T) or false positive (F). 'T' means it is known to be a true SNP/Indel, while a 'F' means it is known not to be a SNP/Indel but the technology used to create the VCF calls it. To annotate the VCF, simply add an INFO field GV with the value T or F.</p>
<h3>The Outputs</h3>
<p>GenotypeAndValidate has two outputs. The <em>truth table</em> and the <em>optional VCF file</em>. The truth table is a 2x2 table correlating what was called in the dataset with the truth of the call (whether it's a true positive or a false positive). The table should look like this: </p>
<table border="1" cellpadding="2" align="center">
<tr>
<th scope="col">
</th>
<th scope="col"> ALT
</th>
<th scope="col"> REF
</th>
<th scope="col"> Predictive Value
</th></tr>
<tr>
<td> <b>called alt</b> </td>
<td> True Positive (TP) </td>
<td> False Positive (FP) </td>
<td> Positive PV
</td></tr>
<tr>
<td> <b>called ref</b> </td>
<td> False Negative (FN) </td>
<td> True Negative (TN) </td>
<td> Negative PV
</td></tr></table>
<p>The <strong>positive predictive value</strong> (PPV)  is the proportion of subjects with <em>positive</em> test results who are correctly diagnose.</p>
<p>The <strong>negative predictive value</strong> (NPV) is the proportion of subjects with a <em>negative</em> test result who are correctly diagnosed.</p>
<p>The optional VCF file will contain only the variants that were called or not called, excluding the ones that were uncovered or didn't pass the filters (-depth). This file is useful if you are trying to compare the PPV and NPV of two different technologies on the exact same sites (so you can compare apples to apples).</p>
<h3>Additional Details</h3>
<ul>
<li>
<p>You should always use -BTI alleles, so that the GATK only looks at the sites on the VCF file, speeds up the process a lot. (this will soon be added as a default gatk engine mode)</p>
</li>
<li>The total number of visited bases may be greater than the number of variants in the original VCF file because of extended indels, as they trigger one call per new insertion or deletion. (i.e. ACTG/- will count as 4 genotyper calls, but it's only one line in the VCF).</li>
</ul>
<h3>Examples</h3>
<p>Genotypes BAM file from new technology using the VCF as a truth dataset:</p>
<pre><code class="pre_md">java \
    -jar /GenomeAnalysisTK.jar \
    -T  GenotypeAndValidate \
    -R human_g1k_v37.fasta \
    -I myNewTechReads.bam \
    -alleles handAnnotatedVCF.vcf \
    -BTI alleles \
    -o gav.vcf</code class="pre_md"></pre>
<p>An annotated VCF example (info field clipped for clarity)</p>
<pre><code class="pre_md">#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  NA12878
1   20568807    .   C   T   0    HapMapHet        AC=1;AF=0.50;AN=2;DP=0;GV=T  GT  0/1
1   22359922    .   T   C   282  WG-CG-HiSeq      AC=2;AF=0.50;GV=T;AN=4;DP=42 GT:AD:DP:GL:GQ  1/0 ./. 0/1:20,22:39:-72.79,-11.75,-67.94:99    ./.
13  102391461   .   G   A   341  Indel;SnpCluster AC=1;GV=F;AF=0.50;AN=2;DP=45 GT:AD:DP:GL:GQ  ./. ./. 0/1:32,13:45:-50.99,-13.56,-112.17:99   ./.
1   175516757   .   C   G   655  SnpCluster,WG    AC=1;AF=0.50;AN=2;GV=F;DP=74 GT:AD:DP:GL:GQ  ./. ./. 0/1:52,22:67:-89.02,-20.20,-191.27:99   ./.</code class="pre_md"></pre>
<p>Using a BAM file as the truth dataset:</p>
<pre><code class="pre_md">java \
    -jar /GenomeAnalysisTK.jar \
    -T  GenotypeAndValidate \
    -R human_g1k_v37.fasta \
    -I myTruthDataset.bam \
    -alleles callsToValidate.vcf \
    -BTI alleles \
    -bt \
    -o gav.vcf</code class="pre_md"></pre>
<p>Example truth table of PacBio reads (BAM) to validate HiSeq annotated dataset (VCF) using the GenotypeAndValidate walker:</p>
<p><img src="http://www.broadinstitute.org/gatk/media/pics/PbGenotypeAndValidate.jpg" alt="PacBio PbGenotypeAndValidate results" /></p>